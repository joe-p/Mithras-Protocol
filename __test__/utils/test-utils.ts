import circom_tester from "circom_tester";
import { AlgorandClient } from "@algorandfoundation/algokit-utils/types/algorand-client";
import { microAlgos, Config } from "@algorandfoundation/algokit-utils";
import algosdk from "algosdk";

Config.configure({
  logger: {
    error: () => {},
    warn: () => {},
    info: () => {},
    verbose: () => {},
    debug: () => {},
  },
});

const wasm_tester = circom_tester.wasm;

export interface CircuitTestConfig {
  circuitPath: string;
  prime?: string;
  recompile?: boolean;
}

export interface MimcInput {
  msgs: [bigint, bigint];
}

export interface MerklePathInput {
  leaf: bigint;
  pathElements: bigint[];
  pathSelectors: number[];
}

export class CircuitTester {
  static async create(config: CircuitTestConfig) {
    return await wasm_tester(config.circuitPath, {
      prime: config.prime || "bls12381",
      recompile: config.recompile ?? true,
    });
  }

  static async createMimcTester() {
    return await this.create({
      circuitPath: "__test__/circuits/mimc.circom",
    });
  }

  static async createMerklePathTester() {
    return await this.create({
      circuitPath: "__test__/circuits/merkle_path_verify.circom",
    });
  }

  static async createMimcSum4Tester() {
    return await this.create({
      circuitPath: "__test__/circuits/mimc_sum4.circom",
    });
  }
}

export class AlgorandTestUtils {
  static createLocalClient(): AlgorandClient {
    return AlgorandClient.defaultLocalNet();
  }

  static async getDispenser(client: AlgorandClient) {
    return await client.account.localNetDispenser();
  }

  static generateRandomAppName(prefix: string): string {
    return `${prefix}-${Math.random().toString(36).substring(2, 15)}`;
  }

  static async fundAccount(
    client: AlgorandClient,
    address: algosdk.Address,
    amount: number,
  ) {
    await client.account.ensureFundedFromEnvironment(
      address,
      microAlgos(amount),
    );
  }
}

export class MimcCalculator {
  private circuit: any;
  private sum4?: any;

  constructor(circuit: any, sum4?: any) {
    this.circuit = circuit;
    this.sum4 = sum4;
  }

  static async create(): Promise<MimcCalculator> {
    const circuit = await CircuitTester.createMimcTester();
    const sum4 = await CircuitTester.createMimcSum4Tester();
    return new MimcCalculator(circuit, sum4);
  }

  async calculateHash(left: bigint, right: bigint): Promise<bigint> {
    const witness = await this.circuit.calculateWitness({
      msgs: [left, right],
    });
    return witness[1];
  }

  async sum4Commit(msgs: [bigint, bigint, bigint, bigint]): Promise<bigint> {
    if (!this.sum4) throw new Error("sum4 circuit not initialized");
    const witness = await this.sum4.calculateWitness({ msgs });
    return witness[1];
  }

  async calculateMerkleRoot(
    leaf: bigint,
    pathElements: bigint[],
    pathSelectors: number[],
  ): Promise<bigint> {
    let currentHash = leaf;

    for (let i = 0; i < pathElements.length; i++) {
      const left = pathSelectors[i] === 0 ? currentHash : pathElements[i];
      const right = pathSelectors[i] === 0 ? pathElements[i] : currentHash;

      currentHash = await this.calculateHash(left, right);
    }

    return currentHash;
  }
}

export class MerkleTestHelpers {
  static createDefaultPathElements(size: number = 32): bigint[] {
    return new Array(size).fill(0n);
  }

  static createDefaultPathSelectors(size: number = 32): number[] {
    return new Array(size).fill(0);
  }

  static bytesToBigInt(bytes: Uint8Array): bigint {
    return BigInt("0x" + Buffer.from(bytes).toString("hex"));
  }

  static bigIntToBytes(value: bigint): Uint8Array {
    const hex = value.toString(16).padStart(64, "0");
    const bytes = new Uint8Array(32);
    for (let i = 0; i < 32; i++) {
      bytes[i] = parseInt(hex.substr(i * 2, 2), 16);
    }
    return bytes;
  }

  static async verifyCircuitWithInput(
    circuit: any,
    input: MerklePathInput,
  ): Promise<void> {
    const witness = await circuit.calculateWitness(input);
    await circuit.checkConstraints(witness);
  }
}

export class TestDataBuilder {
  static createMimcInput(left: bigint, right: bigint): MimcInput {
    return { msgs: [left, right] };
  }

  static createMerklePathInput(
    leaf: bigint,
    pathElements: bigint[] = MerkleTestHelpers.createDefaultPathElements(),
    pathSelectors: number[] = MerkleTestHelpers.createDefaultPathSelectors(),
  ): MerklePathInput {
    return { leaf, pathElements, pathSelectors };
  }

  static createTestLeaf(value: number): Uint8Array {
    const leaf = new Uint8Array(32);
    const bytes = [
      (value >> 24) & 0xff,
      (value >> 16) & 0xff,
      (value >> 8) & 0xff,
      value & 0xff,
    ];
    leaf.set(bytes, 28);
    return leaf;
  }
}
