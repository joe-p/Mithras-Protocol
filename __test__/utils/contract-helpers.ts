import { microAlgos } from "@algorandfoundation/algokit-utils";
import {
  MimcMerkleFactory,
  MimcMerkleClient,
} from "../../contracts/clients/MimcMerkle";
import { MimcTestClient, MimcTestFactory } from "../contracts/clients/MimcTest";
import { AlgorandTestUtils } from "./test-utils";

export class MimcTestHelper {
  static async deployContract(): Promise<MimcTestClient> {
    const algorand = AlgorandTestUtils.createLocalClient();
    const factory = new MimcTestFactory({
      algorand,
      defaultSender: await AlgorandTestUtils.getDispenser(algorand),
    });

    const { appClient } = await factory.deploy({
      appName: AlgorandTestUtils.generateRandomAppName("mimc-test"),
    });

    return appClient;
  }
}

export class MimcMerkleHelper {
  static async deployContract(): Promise<MimcMerkleClient> {
    const algorand = AlgorandTestUtils.createLocalClient();
    const factory = new MimcMerkleFactory({
      algorand,
      defaultSender: await AlgorandTestUtils.getDispenser(algorand),
    });

    const { appClient } = await factory.deploy({
      appName: AlgorandTestUtils.generateRandomAppName("mimc-merkle"),
    });

    await AlgorandTestUtils.fundAccount(
      algorand,
      appClient.appAddress,
      1567900,
    );

    await appClient.send.bootstrap({
      args: {},
      extraFee: microAlgos(256 * 1000),
    });

    return appClient;
  }

  static async addLeaf(
    appClient: MimcMerkleClient,
    leafHash: Uint8Array,
  ): Promise<void> {
    await appClient.send.addLeaf({
      args: { leafHash },
      extraFee: microAlgos(256 * 1000),
    });
  }

  static async getContractState(appClient: MimcMerkleClient) {
    const subtree = await appClient.state.box.subtree();
    const zeroHashes = await appClient.state.box.zeroHashes();

    if (!subtree || !zeroHashes) {
      throw new Error("Failed to get state from contract");
    }

    return { subtree, zeroHashes };
  }

  static async verifyRoot(
    appClient: MimcMerkleClient,
    root: Uint8Array,
  ): Promise<boolean> {
    const { return: isValid } = await appClient.send.isValidRoot({
      args: { root },
    });
    return isValid!;
  }
}

