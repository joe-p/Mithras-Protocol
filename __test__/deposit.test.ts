import { describe, it, beforeAll, expect } from "vitest";
import { CircuitTester, MimcCalculator } from "./utils/test-utils";
import { AppVerifier } from "snarkjs-algorand";
import { AlgorandClient } from "@algorandfoundation/algokit-utils";

describe("Deposit Circuit", () => {
  let circuit: any;
  let mimc: MimcCalculator;

  beforeAll(async () => {
    circuit = await CircuitTester.create({
      circuitPath: "circuits/deposit.circom",
    });
    mimc = await MimcCalculator.create();
  });

  async function sum4Commit(a: bigint, b: bigint, c: bigint, d: bigint) {
    return await mimc.sum4Commit([a, b, c, d]);
  }

  it("computes commitment = MiMC_Sum(spend,null,amount,receiver)", async () => {
    const spending_secret = 111n;
    const nullifier_secret = 222n;
    const amount = 333n;
    const receiver = 444n;

    const expected = await sum4Commit(
      spending_secret,
      nullifier_secret,
      amount,
      receiver,
    );

    const witness = await circuit.calculateWitness({
      amount,
      receiver,
      spending_secret,
      nullifier_secret,
    });
    await circuit.checkConstraints(witness);

    const commitment = witness[1];
    expect(commitment).toBe(expected);
  });

  it("handles zeros", async () => {
    const spending_secret = 0n;
    const nullifier_secret = 0n;
    const amount = 0n;
    const receiver = 0n;

    const expected = await sum4Commit(
      spending_secret,
      nullifier_secret,
      amount,
      receiver,
    );

    const witness = await circuit.calculateWitness({
      amount,
      receiver,
      spending_secret,
      nullifier_secret,
    });
    await circuit.checkConstraints(witness);
    expect(witness[1]).toBe(expected);
  });

  it("verifies on chain", async () => {
    const algorand = AlgorandClient.defaultLocalNet();
    const verifier = new AppVerifier(
      algorand,
      "circuits/deposit_test.zkey",
      "circuits/deposit_js/deposit.wasm",
    );

    await verifier.deploy({
      defaultSender: await algorand.account.localNetDispenser(),
      onUpdate: "append",
    });

    const spending_secret = 111n;
    const nullifier_secret = 222n;
    const amount = 333n;
    const receiver = 444n;

    await verifier.simulateVerification(
      {
        spending_secret,
        nullifier_secret,
        amount,
        receiver,
      },
      { extraOpcodeBudget: 20_000 * 16 },
    );
  });
});
