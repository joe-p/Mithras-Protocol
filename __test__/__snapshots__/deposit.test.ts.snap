// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Deposit Circuit > verifies on chain > deposit verification budget 1`] = `103513`;
